apps/prairielearn/assets/scripts/lib/histogram.ts

```diff
@@ -0,0 +1,125 @@
+import { scaleLinear, axisBottom, axisLeft, select } from 'd3';
+
+export function histogram(
+  selector: Element,
+  data: number[],
+  xgrid: number[],
+  options: {
+    width?: number;
+    height?: number;
+    xmin?: number | 'auto';
+    xmax?: number | 'auto';
+    ymin?: number | 'auto';
+    ymax?: number | 'auto';
+    xlabel?: string;
+    ylabel?: string;
+    xTickLabels?: string[] | 'auto';
+    topMargin?: number;
+    rightMargin?: number;
+    bottomMargin?: number;
+    leftMargin?: number;
+  },
+) {
+  const resolvedOptions = {
+    width: 100,
+    height: 40,
+    xmin: 'auto',
+    xmax: 'auto',
+    ymin: 'auto',
+    ymax: 'auto',
+    xlabel: 'value',
+    ylabel: 'count',
+    xTickLabels: 'auto',
+    topMargin: 10,
+    rightMargin: 20,
+    bottomMargin: 55,
+    leftMargin: 70,
+    ...options,
+  };
+
+  const width = 600 - (resolvedOptions.leftMargin ?? 70) - (resolvedOptions.rightMargin ?? 20);
+  const height = 371 - (resolvedOptions.topMargin ?? 10) - (resolvedOptions.bottomMargin ?? 55);
+
+  const xmin = resolvedOptions.xmin === 'auto' ? Math.min(...xgrid) : resolvedOptions.xmin;
+  const xmax = resolvedOptions.xmax === 'auto' ? Math.max(...xgrid) : resolvedOptions.xmax;
+  if (typeof xmin !== 'number' || typeof xmax !== 'number') {
+    throw new Error('xmin and xmax must be numbers');
+  }
+  const x = scaleLinear().domain([xmin, xmax]).range([0, width]);
+
+  const ymin = resolvedOptions.ymin === 'auto' ? Math.min(...data) : resolvedOptions.ymin;
+  const ymax = resolvedOptions.ymax === 'auto' ? Math.max(...data) : resolvedOptions.ymax;
+  if (typeof ymin !== 'number' || typeof ymax !== 'number') {
+    throw new Error('xmin and xmax must be numbers');
+  }
+  const y = scaleLinear().domain([ymin, ymax]).nice().range([height, 0]);
+
+  let xAxis = axisBottom(x).tickValues(xgrid);
+  if (resolvedOptions.xTickLabels !== 'auto') {
+    xAxis = xAxis.tickFormat((d, i) => resolvedOptions.xTickLabels[i]);
+  }
+  const yAxis = axisLeft(y);
+
+  const xGrid = axisBottom(x)
+    .tickValues(xgrid)
+    .tickSize(-height)
+    .tickFormat(() => '');
+
+  const yGrid = axisLeft(y)
+    .tickSize(-width)
+    .tickFormat(() => '');
+
+  const svg = select(selector)
+    .append('svg')
+    .attr('width', width + (resolvedOptions.leftMargin ?? 0) + (resolvedOptions.rightMargin ?? 0))
+    .attr('height', height + (resolvedOptions.topMargin ?? 0) + (resolvedOptions.bottomMargin ?? 0))
+    .attr('class', 'center-block statsPlot')
+    .append('g')
+    .attr(
+      'transform',
+      'translate(' + resolvedOptions.leftMargin + ',' + resolvedOptions.topMargin + ')',
+    );
+
+  svg
+    .append('g')
+    .attr('class', 'x grid')
+    .attr('transform', 'translate(0,' + height + ')')
```

Comment:
template literals?

Reply:
Updated to template literals in [bd59c6c](https://github.com/PrairieLearn/PrairieLearn/pull/9809/commits/bd59c6c63f9729486ff37c2971ec617424f9ce43).